cpp/src/io/parquet/page_enc.cu

```diff
@@ -1362,26 +1555,46 @@ __device__ void get_extremum(const statistics_val* stats_val,
   }
 
   if (dtype == dtype_string) {
-    *len = stats_val->str_val.length;
-    *val = stats_val->str_val.ptr;
+    return truncate_string(stats_val->str_val, is_min, scratch, truncate_length);
```

Comment:
For some reason I shy away from templates in device code...don't know where I picked up that aversion. I'd need a lot of help to do type dispatching...that makes my head explode.

Reply:
Since only string_view and byte_array_view get truncated, the first template would never be used.  Would it be better to just overload rather than template?

```c++
std::pair<const void*, uint32_t> truncate(string_view const& val, ...) {}
std::pair<const void*, uint32_t> truncate(binary_array_view const& val, ...) {}
```

Reply:
I'm fine either way. The advantage of the template would be that you could blindly call `truncate(val)`, but looking at it, you have to know which type to pull out of the union, so you're not going to save much with that. An overload would be fine, but I'm not sure what it really gains you over `truncate_string` and `truncate_bytes`. This might end up being go full type dispatch with `get_extremum` or do what you have. I'll make an issue as a followup to investigate it and not tie up with review with something like that.

Reply:
I'll leave it be and wait for the dispatch magic :smile:

Reply:
Not what you want, but I reworked `get_extremum`.  There was a lot of cruft leftover from when it was part of `EncodeStatistics`.

Reply:
This is agreeable. I'm not going to force you into building a type dispatcher during a code review. We ask a good deal of submissions, but I like to think we're not sadistic.